# Market Research
## Session 5 - Multiple Linear Regression


In [18]:
# Python imports
import pandas as pd
import numpy as np

## Load data

In [19]:
# load used_cars_data.csv

used_cars_data = pd.read_csv("used_cars_data.csv")

#Display the first five rows of used_cars_data using the .head()

print(used_cars_data.head())

#Use .info() to inspect the DataFrame used_cars_data
print(used_cars_data.info())

   S.No.                              Name    Location  ...  Seats  New_Price  Price
0      0            Maruti Wagon R LXI CNG      Mumbai  ...    5.0        NaN   1.75
1      1  Hyundai Creta 1.6 CRDi SX Option        Pune  ...    5.0        NaN  12.50
2      2                      Honda Jazz V     Chennai  ...    5.0  8.61 Lakh   4.50
3      3                 Maruti Ertiga VDI     Chennai  ...    7.0        NaN   6.00
4      4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  ...    5.0        NaN  17.74

[5 rows x 14 columns]
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7253 entries, 0 to 7252
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   S.No.              7253 non-null   int64  
 1   Name               7253 non-null   object 
 2   Location           7253 non-null   object 
 3   Year               7253 non-null   int64  
 4   Kilometers_Driven  7253 non-null   int64  
 5   Fuel_Type          

In [20]:
# print column names - iterating the columns
for col in used_cars_data.columns:
    print(col)
    
# shape of the data frame
used_cars_data.shape    

#data_types of columns
used_cars_data.dtypes  

S.No.
Name
Location
Year
Kilometers_Driven
Fuel_Type
Transmission
Owner_Type
Mileage
Engine
Power
Seats
New_Price
Price


S.No.                  int64
Name                  object
Location              object
Year                   int64
Kilometers_Driven      int64
Fuel_Type             object
Transmission          object
Owner_Type            object
Mileage               object
Engine                object
Power                 object
Seats                float64
New_Price             object
Price                float64
dtype: object

### Check for Null values

In [21]:
print(used_cars_data.isnull().sum())

S.No.                   0
Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
Mileage                 2
Engine                 46
Power                  46
Seats                  53
New_Price            6247
Price                1234
dtype: int64


From the results above, below columns needs data cleaning for regression for Nulls 
Mileage                 2
Engine                 46
Power                  46
Seats                  53
New_Price            6247
Price                1234


### Data Validation - Total Cells vs Missing %

In [23]:
#Find % of missing data
missing_count = used_cars_data.isnull().sum() #number of missing
total_cells = np.product(used_cars_data.shape) # number of cells (cols x rows)
total_missing = missing_count.sum()
missing_percent = (total_missing*100)/total_cells

print('Total : ', total_cells)
print('Total missing : ', total_missing)
print('Missing Percentage: ', missing_percent, '%')

Total :  101542
Total missing :  7628
Missing Percentage:  7.512162454944752 %


### Descriptive statistics

In [16]:
# Descriptive statistics

used_cars_data.describe(include="all").transpose()

count unique  ...      75%        max
S.No.              7253.0    NaN  ...   5439.0     7252.0
Name                 7253   2041  ...      NaN        NaN
Location             7253     11  ...      NaN        NaN
Year               7253.0    NaN  ...   2016.0     2019.0
Kilometers_Driven  7253.0    NaN  ...  73000.0  6500000.0
Fuel_Type            7253      5  ...      NaN        NaN
Transmission         7253      2  ...      NaN        NaN
Owner_Type           7253      4  ...      NaN        NaN
Mileage              7251    450  ...      NaN        NaN
Engine               7207    150  ...      NaN        NaN
Power                7207    386  ...      NaN        NaN
Seats              7200.0    NaN  ...      5.0       10.0
New_Price            1006    625  ...      NaN        NaN
Price              6019.0    NaN  ...     9.95      160.0
Brand                7253     33  ...      NaN        NaN

[15 rows x 11 columns]

### Data cleaning column by column

- **Name**            
- Location              
- Year                   
- Kilometers_Driven      
- Fuel_Type             
- Transmission          
- Owner_Type            
- Mileage               
- Engine                
- Power                 
- Seats                
- New_Price             
- Price                


In [11]:
used_cars_data['Name']

# First word contains Brand in the Name column, Let's extract brand from it

# Function to extract the brand from the name
def extract_brand(name):
    return name.split()[0]

# Applying the function to the 'name' column to create a 'brand' column
used_cars_data['Brand'] = used_cars_data['Name'].apply(extract_brand)
print(used_cars_data.head())

   S.No.                              Name  ...  Price    Brand
0      0            Maruti Wagon R LXI CNG  ...   1.75   Maruti
1      1  Hyundai Creta 1.6 CRDi SX Option  ...  12.50  Hyundai
2      2                      Honda Jazz V  ...   4.50    Honda
3      3                 Maruti Ertiga VDI  ...   6.00   Maruti
4      4   Audi A4 New 2.0 TDI Multitronic  ...  17.74     Audi

[5 rows x 15 columns]


### Data cleaning column by column

- ~~Name~~            
- **Location**             
- Year                   
- Kilometers_Driven      
- Fuel_Type             
- Transmission          
- Owner_Type            
- Mileage               
- Engine                
- Power                 
- Seats                
- New_Price             
- Price   

In [17]:
used_cars_data['Location']

0           Mumbai
1             Pune
2          Chennai
3          Chennai
4       Coimbatore
           ...    
7248     Hyderabad
7249        Mumbai
7250       Kolkata
7251          Pune
7252         Kochi
Name: Location, Length: 7253, dtype: object

### Data cleaning column by column

- ~~Name~~            
- **Location**             
- Year                   
- Kilometers_Driven      
- Fuel_Type             
- Transmission          
- Owner_Type            
- Mileage               
- Engine                
- Power                 
- Seats                
- New_Price             
- Price   